### ISSUE TITLE
Filter NordVPN servers with jq instead of NordVPN API

### ISSUE THREAD

**Issue by b-kamphorst** [bubuntux/nordlynx#56]
**Is your feature request related to a problem? Please describe.**
I find it hard to appropriately configure the `QUERY` environmental variable. NordVPN API does not seem to match the JSON fieldnames, e.g. `filters\[country_id\]=153` seems random. Since the API is not documented, it is incredibly difficult to find the correct names and values. See also #9, #39, #57, #59, https://github.com/bubuntux/nordlynx/discussions/8, https://github.com/bubuntux/nordlynx/discussions/25.

**Describe the solution you'd like**
I'd like the variable to reflect the naming and values that can be seen at https://api.nordvpn.com/v1/servers/recommendations. This makes it much easier to deduce the correct configuration. At the same time, it is more readable as we can pass names instead of numbers (e.g. `NL` or `Netherlands` instead of `153`).

**Describe alternatives you've considered**
`nordlynx` already uses `jq` to filter relevant information from the NordVPN servers API. My suggestion is to also use `jq` for selecting a server that meets certain properties. For example, [this line](https://github.com/bubuntux/nordlynx/blob/f2d68c9b5d9872a40c2154cd8173dd20bb3cc58a/root/etc/services.d/wireguard/run#L5) could be replaced by 
```
recommendation=$(curl --retry 3 -LsS https://api.nordvpn.com/v1/servers/recommendations | QUERY="${QUERY:-\".[]\"}" jq --raw-output 'first(.[] | .technologies[].name == "Wireguard" and env.QUERY))')
```
A user could pass something like `"QUERY=.locations[].country.name == \"Netherlands\" and .groups[].title == \"P2P\""` to obtain a P2P server in the Netherlands.

_Note: I'm new to jq, so I'm not quite sure if the above example is meaningful. I'd like to hear your thoughts about the proposed change before putting more effort into this._

**Comment by b-kamphorst** [comment]
Also I think that, currently, an incorrect `QUERY` just selects a random server without raising any kind of error or warning. Instead, the suggested approach does not find any server to connect to and the setup should fail.

**Comment by b-kamphorst** [comment]
For backwards compatibility and possibly other reasons, we could also support both `QUERY` and `QUERY_JQ`.

**Comment by yash-fn** [comment]
your title says jp ... did you mean jq?

**Comment by yash-fn** [comment]
Solution to your problem:

1. https://sleeplessbeastie.eu/2019/02/18/how-to-use-public-nordvpn-api/
2. using step 1 find desired hostname (xxx.nordvpn.com) and its associated publickey (similar to your private key)
3. use following docker-compose:
```yml
nordvpn:
    image: bubuntux/nordlynx
    cap_add:
      - NET_ADMIN #required
    environment:
      - PRIVATE_KEY=PRIVATE_KEY
      - END_POINT=subdomain.nordvpn.com:51820
      - PUBLIC_KEY=PUBLIC_KEY
    restart: unless-stopped

```

**Comment by b-kamphorst** [comment]
> your title says jp ... did you mean jq?

I did indeed, fixed it.

> Solution to your problem:
> 1. https://sleeplessbeastie.eu/2019/02/18/how-to-use-public-nordvpn-api/
> 1. using step 1 find desired hostname (xxx.nordvpn.com) and its associated publickey (similar to your private key)
> 1. use following docker-compose:

Thanks for your suggestion, I hadn't considered the `ENDPOINT` variable. The post you shared is exactly the reference that motivated me to propose this feature. Your solution works well as long as the (specs of the) endpoints do not change. I don't suppose they often do. Still, I wonder why we would prefer the nordvpn API over jq?

1. The link you describe documents the jq approach. Such documentation is missing for the nordvpn API. Note how this lies at the root of several community issues in this repo.
1. Even without (updated) documentation, the jq approach is more intuitive than the nordvpn API (e.g. how to filter based on an identifier, how to connect to a specific country (abbreviation/name versus random code)).
1. The jq approach is native to the current codebase. This makes the suggestion fairly simple to implement and does not introduce additional dependencies.

Perhaps a `QUERY_JQ` could replace both `QUERY` and `ENDPOINT`, but of course that would be a breaking change and I can imagine that this is undesired. Having all of them could be confusing as well so that might be a reason to not support `QUERY_JQ`. 

I wonder how you reflect on these thoughts?

**Comment by yash-fn** [comment]
I'm not sure I understand. The approach I gave you allows me to connect to any server I want. This project internally uses same NordVPN api with jq to get the same endpoint and public key data. For simplicity it picks the best suggested server which for most users does the job. Advanced users can always override internal selection by using endpoint variable. I do not speak for developer of this project, but I personally wouldn't want to put in extra work to support every possible API. I also would just pick one that is easiest and stick to that.

**Comment by b-kamphorst** [comment]
Yes, that is the default behavior. The library has a `QUERY` variable and therewith supports server preferences in a presumably simpler way than through providing `PUBLIC_KEY` _and_ `ENDPOINT` (responsibility at library, not at user). The part where we seem to misunderstand (or disagree) is that nordvpn API is or is not "the easiest", or - more specifically - that the undocumented nordvpn API makes it hard to find the appropriate value for `QUERY` and that the combination of nordvpn API and jq may yield a much friendlier way to achieve this support.

Phrased differently: it appears that the `QUERY` variable should make it possible to connect to any server that you want. However, several users have pointed out that it is hard to use. In explaining how people can use it, often to explain how you can retrieve your country code, the community refers to the link that you also shared. If that approach is so helpful, then would it not be easier to start from there?

**Comment by b-kamphorst** [comment]
(That, and the fact that it is explicit about misconfiguration, see https://github.com/bubuntux/nordlynx/issues/56#issuecomment-1067435671.)

**Comment by yash-fn** [comment]
It is not officially documented but somebody has provided unofficial documentation that as of today still works fine:

https://sleeplessbeastie.eu/2019/02/18/how-to-use-public-nordvpn-api/

I followed instructions in link and it was very easy for me to use. Just curl and jq all I need to get proper endpoints and public key. Anyone who has technical ability to use complexity like docker should easily manage curl and jq, no? I'm afraid I still don't understand the concern. It appears you wish for easier API interface, but from my perspective it is already very use to use and I cannot imagine how it could be made easier. Perhaps provide please example of what it is that you imagine should be implemented?

**Comment by b-kamphorst** [comment]
The usage that you suggest is facilitated by nordlynx, but note that that approach is completely agnostic of NordVPN. I think it would work for most wireguard images. I'd say that the key contribution of nordlynx is that it wraps a vanilla wireguard image with nordlynx-specific details.

The most obvious aspect of this it that it finds a suitable NordVPN public key and endpoint for you. That in itself is valuable, but the developer choose to take this one step further. He implemented a `QUERY` variable that can be leveraged to set requirements to the NordVPN server that is connected to. The contents of this variable are passed in the GET request to the NordVPN API. The syntax of this request is NOT documented by the link that you shared and raises many questions in this repository (see my original post).

My suggestion is to add/replace the implementation of `QUERY` so that it is not passed through the GET request, but it is instead used to parse the vanilla JSON as is documented by the link that you provide and are so happy with.

> Perhaps provide please example of what it is that you imagine should be implemented?

Please see #60.  In practice, this replaces `QUERY=filters\[country_id\]=153` by `QUERY_JQ=".locations[].country.name == \"Netherlands\""`. If you want to connect to a server with a certain ID, I wouldn't even know how to do that through `QUERY`.

In conclusion: I feel that using `PUBLIC_KEY` and `ENDPOINT` indicates that you could have used any other Wireguard repo. The benefit of nordlynx is that you can use the `QUERY` variable and I suggest an alternative implementation that I think is easier to use (as enforced by your suggestion of using jq). "Alternative" meaning that it does not directly add any novel functionality, but merely alters the implementation of existing functionality.

If you still don't agree that this _might_ be useful, I wonder what you think of the existing `QUERY` variable?

**Comment by yash-fn** [comment]
Ah yes I think I understand now. Yes that doesn't seem too unreasonable. I agree this could be useful to average users. @bubuntux what are your thoughts? I may experiment with your PR later to see if I can add anything of value.

**Comment by bubuntux** [comment]
yeah having another env variable for jq query seems like a nice addition and probably easier to use than the existing query that relay on undocumented api, PRs are always welcome

**Comment by yash-fn** [comment]
Hey @bubuntux, I'm experimenting with some possibilities in my spare time, but they are all in python. Would you be averse to any solution in python? Or would it be mandatory that any NordVPN Selector method be in bash?

**Comment by bubuntux** [comment]
> Hey @bubuntux, I'm experimenting with some possibilities in my spare time, but they are all in python. Would you be averse to any solution in python? Or would it be mandatory that any NordVPN Selector method be in bash? Seems as though python is not pre-installed inside image so it would make image larger.

I would prefer keep the image smaller, if the value added is significant i would consider it

**Comment by b-kamphorst** [comment]
@yash-fn do you have any thoughts on the proposal?
